### PR TITLE
Compute scheduling key once on creation for jobdb.Job

### PR DESCRIPTION
**Pull request by severinson**
This results in an overall speedup of about 23% when running the simulator for a particular scenario.

Before this change:
```
Measure-Command { .\simulator.exe --clusters .\internal\scheduler\simulator\testdata\clusters\cpu_1_1_100.yaml --workloads '.\internal\scheduler\simulator\testdata\workloads\small_big\*.yaml' --configs .\internal\scheduler\simulator\testdata\configs\defaultSchedulingConfig.yaml --logInterval 10000 }
...
Seconds           : 21
Milliseconds      : 70
Ticks             : 210707754
TotalDays         : 0.000243874715277778
TotalHours        : 0.00585299316666667
TotalMinutes      : 0.35117959
TotalSeconds      : 21.0707754
TotalMilliseconds : 21070.7754
```

After this change:
```
Measure-Command { .\simulator.exe --clusters .\internal\scheduler\simulator\testdata\clusters\cpu_1_1_100.yaml --workloads '.\internal\scheduler\simulator\testdata\workloads\small_big\*.yaml' --configs .\internal\scheduler\simulator\testdata\configs\defaultSchedulingConfig.yaml --logInterval 10000 }
...
Seconds           : 17
Milliseconds      : 69
Ticks             : 170699021
TotalDays         : 0.000197568311342593
TotalHours        : 0.00474163947222222
TotalMinutes      : 0.284498368333333
TotalSeconds      : 17.0699021
TotalMilliseconds : 17069.9021
```

### CODE REVIEW

File: internal/scheduler/jobdb/job.go [hunk 1]
```diff
@@ -58,39 +60,6 @@ func EmptyJob(id string) *Job {
 	return &Job{id: id, runsById: map[uuid.UUID]*JobRun{}}
 }
 
-// NewJob creates a new scheduler job
-func NewJob(
```
Comment (severinson) [inline internal/scheduler/jobdb/job.go:62]: The biggest change in this PR is that it's now necessary to have a jobDb to create jobs. The benefit of this approach is that the jobDb can store various defaults and cached values, which may be useful in the future. This is like what we do with the nodeDb, where the nodeDb is responsible for creating nodes.

File: internal/scheduler/jobdb/jobdb.go [hunk 2]
```diff
@@ -15,25 +17,68 @@ var (
...
+		jobsByQueue:            map[string]immutable.SortedSet[*Job]{},
+		queuedJobsByTtl:        &emptyQueuedJobsByTtl,
+		schedulingKeyGenerator: schedulerobjects.NewSchedulingKeyGenerator(),
+		copyMutex:              sync.Mutex{},
+	}
+}
+
+// NewJob creates a new scheduler job.
+// The new job is not automatically inserted into the jobDb; call jobDb.Upsert to upsert it.
+func (jobDb *JobDb) NewJob(
```
Comment (severinson) [inline internal/scheduler/jobdb/jobdb.go:44]: This the same as the old `NewJob`, except it also computes the `SchedulingKey`.

File: internal/scheduler/queue_scheduler.go [hunk 3]
```diff
@@ -210,7 +210,10 @@ func (it *QueuedGangIterator) Peek() (*schedulercontext.GangSchedulingContext, e
 
 		// Skip this job if it's known to be unschedulable.
 		if len(it.schedulingContext.UnfeasibleSchedulingKeys) > 0 {
-			schedulingKey := it.schedulingContext.SchedulingKeyFromLegacySchedulerJob(job)
+			schedulingKey, ok := job.GetSchedulingKey()
+			if !ok {
+				schedulingKey = it.schedulingContext.SchedulingKeyFromLegacySchedulerJob(job)
```
Comment (severinson) [inline internal/scheduler/queue_scheduler.go:215]: This fallback is necessary to handle api.Job, which doesn't have a pre-computed schedulingKey associated with it.

File: internal/scheduler/scheduler.go [hunk 4]
```diff
@@ -339,12 +339,10 @@ func (s *Scheduler) syncState(ctx *armadacontext.Context) ([]*jobdb.Job, error)
 	}
 
 	jobsToUpdate := maps.Values(jobsToUpdateById)
-	err = s.jobDb.Upsert(txn, jobsToUpdate)
-	if err != nil {
+	if err := s.jobDb.Upsert(txn, jobsToUpdate); err != nil {
```
Comment (severinson) [inline internal/scheduler/scheduler.go:342]: Drive-by change.

**Review by d80tb7 — DISMISSED** [review]

File: internal/scheduler/context/context.go [hunk 5]
```diff
@@ -225,7 +225,6 @@ func (sctx *SchedulingContext) ReportString(verbosity int32) string {
 
 func (sctx *SchedulingContext) AddGangSchedulingContext(gctx *GangSchedulingContext) (bool, error) {
 	allJobsEvictedInThisRound := true
-	allJobsSuccessful := true
```
Comment (severinson) [inline internal/scheduler/context/context.go:228]: Wasn't used.

File: internal/scheduler/gang_scheduler.go [hunk 6]
```diff
@@ -60,29 +83,19 @@ func (sch *GangScheduler) updateGangSchedulingContextOnFailure(gctx *schedulerco
...
-
-	return
-}
-
-func (sch *GangScheduler) updateGangSchedulingContextOnSuccess(gctx *schedulercontext.GangSchedulingContext, gangAddedToSchedulingContext bool) (err error) {
-	if gangAddedToSchedulingContext {
-		jobs := util.Map(gctx.JobSchedulingContexts, func(jctx *schedulercontext.JobSchedulingContext) interfaces.LegacySchedulerJob { return jctx.Job })
-		if _, err = sch.schedulingContext.EvictGang(jobs); err != nil {
-			return
+		if schedulingKey != schedulerobjects.EmptySchedulingKey {
```
Comment (severinson) [inline internal/scheduler/gang_scheduler.go:90]: Guard against future bugs should we forget to compute the key.
